services/src/machine_learning/xgb_config_builder.rs

```diff
@@ -0,0 +1,1057 @@
+    }
+}
+
+impl<T: PartialOrd + Display> Interval<T> {
+    fn new(min: T, min_inclusion: Inclusion, max: T, max_inclusion: Inclusion) -> Self {
+        Interval {
+            min,
+            min_inclusion,
+            max,
+            max_inclusion,
+        }
+    }
+
+    fn new_open_open(min: T, max: T) -> Self {
+        Interval::new(min, Inclusion::Open, max, Inclusion::Open)
+    }
+
+    fn new_open_closed(min: T, max: T) -> Self {
+        Interval::new(min, Inclusion::Open, max, Inclusion::Closed)
+    }
+
+    fn new_closed_closed(min: T, max: T) -> Self {
+        Interval::new(min, Inclusion::Closed, max, Inclusion::Closed)
+    }
+
+    fn contains(&self, val: &T) -> bool {
+        let min = match self.min_inclusion {
+            Inclusion::Closed => {
+                matches!(
+                    val.partial_cmp(&self.min),
+                    Some(Ordering::Equal | Ordering::Greater)
+                )
+            }
+            Inclusion::Open => {
+                matches!(val.partial_cmp(&self.min), Some(Ordering::Greater))
+            }
+        };
+
+        let max = match self.max_inclusion {
+            Inclusion::Closed => {
+                matches!(
+                    val.partial_cmp(&self.max),
+                    Some(Ordering::Equal | Ordering::Less)
+                )
+            }
+            Inclusion::Open => matches!(val.partial_cmp(&self.max), Some(Ordering::Less)),
+        };
+
+        min & max
+    }
+
+    fn validate(&self, val: &Option<T>, name: &str) -> Result<(), String> {
+        match val {
+            Some(ref val) => {
+                if self.contains(val) {
+                    Ok(())
+                } else {
+                    Err(format!(
+                        "Invalid value for '{}' parameter, {} is not in range {}.",
+                        name, &val, self
+                    ))
+                }
+            }
+            None => Ok(()),
+        }
+    }
+}
+
+/// Type of booster to use when training a [Booster](../struct.Booster.html) model.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub enum BoosterType {
+    Tree(TreeBoosterParameters),
+
+    Linear(LinearBoosterParameters),
+}
+
+impl Default for BoosterType {
+    fn default() -> Self {
+        BoosterType::Tree(TreeBoosterParameters::default())
+    }
+}
+
+impl BoosterType {
+    fn as_string_pairs(&self) -> Vec<(String, String)> {
+        match *self {
+            BoosterType::Tree(ref p) => p.as_string_pairs(),
+            BoosterType::Linear(ref p) => p.as_string_pairs(),
+        }
+    }
+}
+
+impl<'a> From<&'a str> for BoosterType {
+    fn from(s: &'a str) -> Self {
+        match s {
+            "gbtree" => BoosterType::Tree(TreeBoosterParameters::default()),
+            "gblinear" => BoosterType::Linear(LinearBoosterParameters::default()),
+            // TODO: better error handling
+            _ => panic!("Unknown or unsupported booster type: {s}"),
```

Comment:
Implement TryFrom instead of From and return an error

Reply:
a74e3309

--- 

Commit MSG:

refactor: remove all panic and unwrap calls from xgb config builder